pkg/tecdsa/signing/states.go

```diff
@@ -2,126 +2,73 @@ package signing
 
 import (
 	"context"
-
 	"github.com/keep-network/keep-core/pkg/net"
+	"github.com/keep-network/keep-core/pkg/protocol/faststate"
 	"github.com/keep-network/keep-core/pkg/protocol/group"
-	"github.com/keep-network/keep-core/pkg/protocol/state"
+	"sync"
 )
 
-const (
-	silentStateDelayBlocks  = 0
-	silentStateActiveBlocks = 0
-
-	ephemeralKeyPairStateDelayBlocks  = 1
-	ephemeralKeyPairStateActiveBlocks = 5
-
-	tssRoundOneStateDelayBlocks  = 1
-	tssRoundOneStateActiveBlocks = 11
-
-	tssRoundTwoStateDelayBlocks  = 1
-	tssRoundTwoStateActiveBlocks = 6
-
-	tssRoundThreeStateDelayBlocks  = 1
-	tssRoundThreeStateActiveBlocks = 5
-
-	tssRoundFourStateDelayBlocks  = 1
-	tssRoundFourStateActiveBlocks = 5
-
-	tssRoundFiveStateDelayBlocks  = 1
-	tssRoundFiveStateActiveBlocks = 5
-
-	tssRoundSixStateDelayBlocks  = 1
-	tssRoundSixStateActiveBlocks = 5
-
-	tssRoundSevenStateDelayBlocks  = 1
-	tssRoundSevenStateActiveBlocks = 5
-
-	tssRoundEightStateDelayBlocks  = 1
-	tssRoundEightStateActiveBlocks = 5
-
-	tssRoundNineStateDelayBlocks  = 1
-	tssRoundNineStateActiveBlocks = 5
-
-	finalizationStateDelayBlocks  = 1
-	finalizationStateActiveBlocks = 5
-)
-
-// ProtocolBlocks returns the total number of blocks it takes to execute
-// all the required work defined by the signing protocol.
-func ProtocolBlocks() uint64 {
-	return ephemeralKeyPairStateDelayBlocks +
-		ephemeralKeyPairStateActiveBlocks +
-		tssRoundOneStateDelayBlocks +
-		tssRoundOneStateActiveBlocks +
-		tssRoundTwoStateDelayBlocks +
-		tssRoundTwoStateActiveBlocks +
-		tssRoundThreeStateDelayBlocks +
-		tssRoundThreeStateActiveBlocks +
-		tssRoundFourStateDelayBlocks +
-		tssRoundFourStateActiveBlocks +
-		tssRoundFiveStateDelayBlocks +
-		tssRoundFiveStateActiveBlocks +
-		tssRoundSixStateDelayBlocks +
-		tssRoundSixStateActiveBlocks +
-		tssRoundSevenStateDelayBlocks +
-		tssRoundSevenStateActiveBlocks +
-		tssRoundEightStateDelayBlocks +
-		tssRoundEightStateActiveBlocks +
-		tssRoundNineStateDelayBlocks +
-		tssRoundNineStateActiveBlocks +
-		finalizationStateDelayBlocks +
-		finalizationStateActiveBlocks
-}
-
 // ephemeralKeyPairGenerationState is the state during which members broadcast
 // public ephemeral keys generated for other members of the group.
 // `ephemeralPublicKeyMessage`s are valid in this state.
 type ephemeralKeyPairGenerationState struct {
+	*faststate.BaseState
+
+	action *stateAction
+
 	channel net.BroadcastChannel
 	member  *ephemeralKeyPairGeneratingMember
-
-	phaseMessages []*ephemeralPublicKeyMessage
 }
 
-func (ekpgs *ephemeralKeyPairGenerationState) DelayBlocks() uint64 {
-	return ephemeralKeyPairStateDelayBlocks
-}
+func (ekpgs *ephemeralKeyPairGenerationState) Initiate(ctx context.Context) error {
+	ekpgs.action.run(func() error {
```

Comment:
Why do we need to wrap the call to `generateEphemeralKeyPair()` with `ekpgs.action.run`? The state machine should guarantee `Initiate` is called only one time.

Reply:
This is not about the number of calls to `Initiate`. The reasoning is as follows:
- we need to wait until all ephemeral keys are generated and the state message is sent before we proceed to the next state
- currently, the state machine implementation guarantees that `Initiate` is called before `CanTransition` but this is an implementation detail that can change in the future so we cannot consider it as a strong guarantee. Because of that, we should check whether the state completed its computations and sent its message before proceeding in the state implementation as well. We don't want to check state internals (i.e. the number of generated keys) as this is not the right abstraction level for that. Additionally, that check does not guarantee that `channel.Send` was called
- `Initiate` should return as fast as possible as the message buffer can become filled meanwhile. Running an async action matches that requirement
- Almost all states of signing perform async computations so using this pattern for `ephemeralKeyPairGenerationState` is justified as we have consistency across all states

Reply:
It seems to me that this is a problem with a state machine and not something the state implementation should worry about.

What do you think about:
- Making `AsyncState.Initiate` asynchronous in the state machine.
- Making `AsyncState.Initiate` return a channel when the computations are done.
- Probably getting rid of `stateAction` structure.

I can try experimenting with this approach, if it works fine, we can open another PR that will be merged immediately after this one.

Reply:
I think that will make the state machine more complicated as we need to examine the state of computations triggered by `Initiate` in `CanTransition`. There is also a risk that we'll specialize the machine too much and it will not fit other possible use cases well enough. Of course, we can always drive some experiments and see.

Reply:
Agreed, this will move more logic to the state machine but right now this logic is outside of the state machine and complicates state implementation. We do not necessarily need to modify the machine, we could also move some of this logic to the base async state. Anyway, something to experiment with.

Some loose thoughts here:
1. `stateAction` ensures the given action can be triggered only one time. This is something the Initiate interface should be clear about and ensure. The state implementation should just trust this assumption.
 
2. In the async state machine we gather all messages received, also those from the future. Blocking `Initiate` makes sense for the sync state machine, where we may want to initiate a state before we are ready to accept messages. For the async state machine, we want to accept messages all the time and blocking Initiate is not really needed.

3. On the abstract level, the state machine should not be asking the state if it `CanTransition` if `Initiate` of the state has not been completed yet, no matter if `Initiate` is blocking or not. This is happening when `Initiate` is blocking, but requires some more logic from the state if `Initiate` logic is async.

Reply:
Addressed in https://github.com/keep-network/keep-core/pull/3385.